File: 专题/状态DP/股票问题/121.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int empty[n+1], exist[n+1];
        empty[0] = 0;
        exist[0] = INT_MIN;
        for (int i = 0; i < n; ++i) {
            empty[i+1] = max(empty[i+0], exist[i] + prices[i]);
            exist[i+1] = max(exist[i+0], - prices[i]);
        }
        return empty[n];
    }
    int maxProfit2(vector<int>& prices) {
        int n = prices.size();
        int empty = 0, exist = INT_MIN;   
        for (int i = 0; i < n; ++i) {
            empty = max(empty, exist + prices[i]);
            exist = max(exist, - prices[i]);
        }
        return empty;
    }
};

File: 专题/状态DP/股票问题/121.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        if not n: return 0
        dp = [ [0] * (n+1) for i in range(2)]
        dp[1][0] = -10000000000  # 这里代表最大
        for i in range(n):
            dp[0][i+1] = max(dp[0][i], dp[1][i] + prices[i])
            dp[1][i+1] = max(dp[1][i], -prices[i])
        return dp[0][n] 
    
    def maxProfit2(self, prices: List[int]) -> int:
        n = len(prices)
        if not n: return 0
        empty = 0
        exist = -100000000000
        for i in range(n):
            empty = max(empty, exist + prices[i])
            exist = max(exist, -prices[i])
        return empty